### DISCUSSION TITLE
作者是否有计划支持 SM9

### DISCUSSION THREAD

**Discussion by wzhvictor** [emmansun/gmsm#50] [General]

**Comment by emmansun** [comment]
我最近也学习了一下

1. 国内SM9的使用场景有哪些？
虽然在TLCP中有说支持SM9, 但实际使用、实现者看起来不多。还有就是奥联自己的安全邮箱。不知道还有哪些其它场景，譬如国内区块链应用？

2. 安全担忧？
SM9使用BN256曲线，虽然其与ISO BN256曲线参数t不同，但安全级别应该可参考，差不多是100bits左右。具体可参考[双线性对友好的曲线](https://datatracker.ietf.org/doc/pdf/draft-irtf-cfrg-pairing-friendly-curves-07.pdf) 以及 https://docs.google.com/spreadsheets/d/1t6ibmx_BqCxyovP18RfGEUUtKrl2d5MD/edit#gid=1450887046。

3. 性能
SM9的性能相对差，涉及2条曲线、三个环域上的运算。

4. 参考
由于SM9的曲线也采用BN256, 目前参考比较多的：
go语言实现： https://github.com/cloudflare/bn256
c语言实现：https://crypto.stanford.edu/pbc

5. 现有实现
c语言有gmssl中的参考实现；
golang语言有https://github.com/xlcetc/cryptogm， 其也是用https://github.com/cloudflare/bn256，修改参数实现的。

6. 会不会在本库中实现？
目前还在学习评估阶段

**Reply by wzhvictor** [reply]
感谢回复，在我看来，SM9 主要用于替代 PKI/CA 体系，在 KMS 中可以通过 KGC 给指定应用签发证书，而 KGC 无需托管私钥

**Reply by emmansun** [reply]
我想替换PKI体系的话，目前来看还不成熟，“可信”KGC本身也是一个问题。“在 KMS 中可以通过 KGC 给指定应用签发证书，而 KGC 无需托管私钥”，不是很理解，PKI是不托管私钥的，用户私钥通常在用户端生成，反而KGC隐式托管私钥。

**Reply by wzhvictor** [reply]
是的，所以 KGC 这套方案也仅在内网尝试，由一些其他基础设施保证应用的安全性，KGC 只需信任对应的标识信息即可推导出私钥，替代由 CA 来生成公私钥，之前由于应用可能会丢失私钥证书，或者不便放在代码仓库中，所以私钥也由 KMS 来保存

**Comment by emmansun** [comment]
SM9 key 生成，签名、验签，密钥封装、解封，加解密（XOR）的初始版本已经就绪，欢迎试用！暂未实现的包括密钥交换，和加解密(分组加解密)法。
